File: Easy/509. Fibonacci Number/solution (4).py
# 509. Fibonacci Number

# Runtime: 28 ms, faster than 85.55% of Python3 online submissions for Fibonacci Number.

# Memory Usage: 14.2 MB, less than 41.38% of Python3 online submissions for Fibonacci Number.


class Solution:
    # Iterative Bottom-Up Approach
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        else:
            curr = 0
            prev1, prev2 = 1, 0
            for i in range(2, n + 1):
                curr = prev1 + prev2
                prev1, prev2 = curr, prev1
            return curr

File: Easy/1137. N-th Tribonacci Number/solution (4).cpp
// 1137. N-th Tribonacci Number


class Solution {
public:
    int tribonacci(int n) {
        if (n == 0) {
            return 0;
        } else if (n <= 2) {
            return 1;
        }

        vector<int> vals(n + 1);
        vals[1] = 1;
        vals[2] = 1;
        for (int i = 3; i <= n; i++) {
            vals[i] = vals[i - 1] + vals[i - 2] + vals[i - 3];
        }

        return vals.back();
    }
};